airbyte-integrations/connectors/source-square/source_square/source.py

```diff
@@ -0,0 +1,426 @@
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+import json
+from abc import ABC
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Tuple
+
+import pendulum
+import requests
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.http import HttpStream
+from airbyte_cdk.sources.streams.http.auth import TokenAuthenticator
+
+
+class SquareStream(HttpStream, ABC):
+    def __init__(self, is_sandbox: bool, api_version: str, start_date: str, include_deleted_objects: bool, **kwargs):
+        super().__init__(**kwargs)
+        self.is_sandbox = is_sandbox
+        self.api_version = api_version
+        # Converting users ISO 8601 format (YYYY-MM-DD) to RFC 3339 (2021-06-14T13:47:56.799Z)
+        # Because this standard is used by square in 'updated_at' records field
+        self.start_date = "{}".format(pendulum.parse(start_date))
+        self.include_deleted_objects = include_deleted_objects
+
+    data_field = None
+    primary_key = "id"
+    items_per_page_limit = 100
+
+    @property
+    def url_base(self) -> str:
+        return "https://connect.squareup{}.com/v2/".format("sandbox" if self.is_sandbox else "")
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        next_page_cursor = response.json().get("cursor", False)
+        if next_page_cursor:
+            return {"cursor": next_page_cursor}
+
+    def request_headers(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> Mapping[str, Any]:
+        return {"Square-Version": self.api_version, "Content-Type": "application/json"}
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        json_response = response.json()
+        records = json_response.get(self.data_field, []) if self.data_field is not None else json_response
+        yield from records
+
+    def _send_request(self, request: requests.PreparedRequest) -> requests.Response:
+        try:
+            return super()._send_request(request)
+        except requests.exceptions.HTTPError as e:
+            if e.response.content:
+                content = json.loads(e.response.content.decode())
+                if content and "errors" in content:
+                    raise SquareException(str(content["errors"]))
```

Comment:
why do you need this error handling ? maybe it's easier to just throw an exception ?

Reply:
in the 'errors' section the API returns the exact field and cause of the error in request.
Throwing just the exception is less informative (just the status code and "Bad request" as the reason)
But in case of this code we have something like below in the log.
`{
  "errors": [
    {
      "code": "BAD_REQUEST",
      "detail": "Must provide at least 1 location_id.",
      "field": "location_ids",
      "category": "INVALID_REQUEST_ERROR"
    }
  ]
}`

Reply:
can you just skip request if there no locations ?